diff --git a/src/keys.rs b/src/keys.rs
index 4eab2c5..fb9a0e7 100644
--- a/src/keys.rs
+++ b/src/keys.rs
@@ -156,6 +156,17 @@ impl SpendAuthorizingKey {
         self.0.randomize(randomizer)
     }

+    /// Randomizes this spend authorizing key with the given `randomizer`, deriving
+    /// the randomized verification key using Ledger SDK Pallas primitives.
+    ///
+    /// The resulting key can be used to actually sign a spend.
+    pub fn randomize_ledger(
+        &self,
+        randomizer: &pallas::Scalar,
+    ) -> Result<redpallas::SigningKey<SpendAuth>, ledger_zcash_crypto::Error> {
+        self.0.randomize_ledger(randomizer)
+    }
+
     /// Creates a RedPallas spend authorization signing key from the given ledger signing key.
     pub fn ledger_try_from(sk: &SpendingKey) -> Result<Self, ledger_zcash_crypto::Error> {
         let ask = Self::ledger_derive_inner(sk)?;
diff --git a/src/primitives/redpallas.rs b/src/primitives/redpallas.rs
index c04f7cd..20b94ed 100644
--- a/src/primitives/redpallas.rs
+++ b/src/primitives/redpallas.rs
@@ -62,6 +62,26 @@ impl SigningKey<SpendAuth> {
     pub fn randomize(&self, randomizer: &pallas::Scalar) -> Self {
         SigningKey(self.0.randomize(randomizer))
     }
+
+    /// Randomizes this signing key with the given `randomizer`, deriving the
+    /// randomized verification key using Ledger SDK Pallas primitives.
+    ///
+    /// Randomization is only supported for `SpendAuth` keys.
+    pub fn randomize_ledger(
+        &self,
+        randomizer: &pallas::Scalar,
+    ) -> Result<Self, ledger_zcash_crypto::Error> {
+        self.0.randomize_ledger(randomizer).map(SigningKey)
+    }
+
+    /// Creates a SpendAuth signature using Ledger SDK RedPallas primitives.
+    pub fn sign_ledger<R: RngCore + CryptoRng>(
+        &self,
+        rng: R,
+        msg: &[u8],
+    ) -> Result<Signature<SpendAuth>, ledger_zcash_crypto::Error> {
+        self.0.sign_ledger(rng, msg).map(Signature)
+    }
 }

 impl<T: SigType> SigningKey<T> {
